### PR TITLE
Refactored PKI Keeper to reduce code duplication 

### PR DESCRIPTION
**Pull request by Artemkaaas**

### CODE REVIEW

File: x/pki/tests/test-design.md [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+### Propose adding of DA root certificate
+
+Indexes to check:
+
+* Present:
+    * `ProposedCertificate`
+    * `UniqueCertificate`
+* Missing:
+    * `RejectedCertificate`
```
Comment (ashcherbakov) [inline x/pki/tests/test-design.md:11]: What about Revoked?

File: x/pki/tests/test-design.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+Test cases:
+
+* Positive:
+    * Propose adding of DA root certificate: `TestHandler_ProposeAddDaRootCert`
+    * Propose adding of previously rejected DA root certificate: ?
+    * Propose adding of DA root certificate with same Subject/SKID as existing Approved certificate but different Serial
+      Number: `TestHandler_ProposeAddX509RootCert_ForDifferentSerialNumber` (need to rewrite)
+* Negative:
+    * TBD
```
Comment (ashcherbakov) [inline x/pki/tests/test-design.md:23]: What does negative mean here?

File: x/pki/keeper/grpc_query_revoked_certificates.go [hunk 3]
```diff
@@ -57,3 +57,17 @@ func (k Keeper) RevokedCertificates(c context.Context, req *types.QueryGetRevoke
 
 	return &types.QueryGetRevokedCertificatesResponse{RevokedCertificates: val}, nil
 }
+
+// IsRevokedCertificatePresent Check if the Revoked Certificate is present in the store.
```
[the remark this answers is not in the document]
Comment (Artemkaaas) [reply x/pki/keeper/grpc_query_revoked_certificates.go:61]: No.  Thanks for pointing.

File: x/pki/tests/test-design.md (hunk 1 above)
Comment (Artemkaaas) [reply x/pki/tests/test-design.md:11]: Is it possible to restore revoked DA root certificate?
Right now I see that none of `ProposeAddX509RootCert` / `ApproveAddX509RootCert` message handlers interacts with revoked certificate indexes.

File: x/pki/keeper/certificate_helpers.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+) {
+	RemoveCertFromList(issuer, serialNumber, &certificates.Certs)
+
+	if len(certificates.Certs) == 0 {
+		k.RemoveDaCertificate(ctx, subject, subjectKeyID, isRoot)
+	} else {
+		k.RemoveAllCertificatesBySerialNumber(ctx, subject, subjectKeyID, serialNumber)
+		k.RemoveAllCertificatesBySubjectKeyIDBySerialNumber(ctx, subject, subjectKeyID, serialNumber)
+		k.RemoveApprovedCertificatesBySerialNumber(ctx, subject, subjectKeyID, serialNumber)
+		k.RemoveApprovedCertificatesBySubjectKeyIDBySerialNumber(ctx, subject, subjectKeyID, serialNumber)
```
[the remark this answers is not in the document]
Comment (ashcherbakov) [reply x/pki/keeper/certificate_helpers.go:172]: OK, but what if we have exactly 1 cert with the given subject+SKID+SN, and we remove it? We should remove it from the index by subject as well... Do we have a corresponding test for this?

[the remark this answers is not in the document]
Comment (Artemkaaas) [reply x/pki/keeper/certificate_helpers.go:172]: This case will be handled by `if` logic.  
`RemoveCertFromList` mutates `certificates.Certs` list, deleting cert by serial number.  
Probably we should rewrite it to be more transparent.  

```
	RemoveCertFromList(issuer, serialNumber, &certificates.Certs)

	if len(certificates.Certs) == 0 {
		k.RemoveDaCertificate(ctx, subject, subjectKeyID, isRoot)
	} else {
	   .....
	```

File: x/pki/tests/test-design.md (hunk 1 above)
Comment (Artemkaaas) [reply x/pki/tests/test-design.md:11]: Following this logic we also should check all noc indexes and so on.

File: x/pki/tests/test-design.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    * `ProposedCertificate`
+
+Test cases:
+
+* Positive:
+    * Add DA intermediate certificate: `TestHandler_AddDaIntermediateCert`
+* Negative:
+    * TBD
+
+## [Revoke DA Root](./handler_revoke_paa_cert_test.go)
```
Comment (ashcherbakov) [inline x/pki/tests/test-design.md:83]: Can you please analyze tests for remove/revoke by SerialNumber (at least simple cases)?

**Review by ashcherbakov — APPROVED** [review]

**Pull request merged**